### PR TITLE
Add package for serenity and its scine wrapper

### PR DESCRIPTION
**Pull request by awvwgk**
Waiting for #34969.

### CODE REVIEW

**Comment by ltalirz** [comment]
Great, thanks a lot @awvwgk !

Once this is merged, perhaps one could try upstreaming some of the edits to the cmake files of serenity.

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            "target_link_libraries(xc INTERFACE ${pc_libxc_LINK_LIBRARIES})\n"
+            "target_include_directories(xc INTERFACE ${pc_libxc_INCLUDE_DIRS})\n"
+            "endif()",
+            "cmake/ImportLibxc.cmake",
+            string=True,
+        )
+
+        filter_file(
+            "function(import_pybind11)",
+            "function(import_pybind11)\n" "find_package(pybind11)" "cmake/ImportPybind11.cmake",
```
Comment (ltalirz) [inline var/spack/repos/builtin/packages/serenity/package.py]:
```suggestion
            "function(import_pybind11)\n" "find_package(pybind11)", 
            "cmake/ImportPybind11.cmake",
```

**Comment by awvwgk** [comment]
> Once this is merged, perhaps one could try upstreaming some of the edits to the cmake files of serenity.

Worth a try, those patches are mostly to devendor the dependencies and would be generally useful in the upstream package as well.

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            "add_library(xcfun INTERFACE IMPORTED)\n"
+            "target_link_libraries(xcfun INTERFACE XCFun::xcfun)\n",
+            "cmake/ImportXCFun.cmake",
+            string=True,
+        )
+
+    def cmake_args(self):
+        args = [
+            self.define("SERENITY_BUILD_TESTS", self.run_tests),
+            self.define("SERENITY_BUILD_PYTHON_BINDINGS", "+python" in self.spec),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py]:
```suggestion
            self.define_from_variant("SERENITY_BUILD_PYTHON_BINDINGS", "python"),
```

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            self.define("SERENITY_MARCH", ""),
+            self.define("SERENITY_PREFER_XCFUN", False),
+            self.define("SERENITY_USE_XCFUN", True),
+            self.define("SERENITY_USE_LIBXC", True),
+            self.define(
+                "SERENITY_USE_INTEL_MKL", self.spec["lapack"].libs.names[0].startswith("mkl")
+            ),
+            self.define("SERENITY_USE_LAPACK", "+lapack" in self.spec),
+            self.define("SERENITY_USE_BLAS", "+blas" in self.spec),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py]:
```suggestion
            self.define_from_variant("SERENITY_USE_LAPACK", "lapack"),
            self.define_from_variant("SERENITY_USE_BLAS", "blas"),
```

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            "target_link_libraries(xcfun INTERFACE XCFun::xcfun)\n",
+            "cmake/ImportXCFun.cmake",
+            string=True,
+        )
+
+    def cmake_args(self):
+        args = [
+            self.define("SERENITY_BUILD_TESTS", self.run_tests),
+            self.define("SERENITY_BUILD_PYTHON_BINDINGS", "+python" in self.spec),
+            self.define("SERENITY_MARCH", ""),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py:120]: What does this setting imply? Note that we'll inject `-march` or similar flags via the compiler wrapper.

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    depends_on("cmake@3.12:", type="build")
+    depends_on("boost")
+    depends_on("eigen@3:")
+    depends_on("googletest@1.8.1:")
+    depends_on("hdf5@1.10.1:")
+    depends_on("lapack", when="+lapack")
+    depends_on("libecpint")
+    depends_on("libxc@5.0.0")
+    depends_on("pkgconfig", type="build")
+    depends_on("python@3.6:", when="+python", type=("build", "run"))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py:33]: I think this is missing:
```python
extends("python", when="+python")
```

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    variant("blas", default=True, description="Use BLAS library with Eigen")
+    variant("lapack", default=True, description="Use Lapack library with Eigen")
+    variant("python", default=False, description="Build Python bindings")
+
+    depends_on("blas", when="+blas")
+    depends_on("cmake@3.12:", type="build")
+    depends_on("boost")
+    depends_on("eigen@3:")
+    depends_on("googletest@1.8.1:")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py]: Maybe:
```suggestion
    depends_on("googletest@1.8.1:", type="test")
```
?

File: var/spack/repos/builtin/packages/serenity/package.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            "cmake/DownloadProject.cmake",
+            string=True,
+        )
+
+        filter_file(
+            "find_package(GTest 1.8.1 QUIET)",
+            "find_package(GTest REQUIRED)",
+            "cmake/ImportGTest.cmake",
+            string=True,
+        )
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/serenity/package.py]: Is this really always required?

File: var/spack/repos/builtin/packages/scine-serenity/package.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            "find_package(OpenMP REQUIRED)\n"
+            "target_compile_options(serenity INTERFACE ${OpenMP_CXX_FLAGS})\n",
+            "CMakeLists.txt",
+            string=True,
+        )
+
+    def cmake_args(self):
+        args = [
+            self.define("SCINE_BUILD_TESTS", self.run_tests),
+            self.define("SCINE_BUILD_PYTHON_BINDINGS", "+python" in self.spec),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/scine-serenity/package.py]:
```suggestion
            self.define_from_variant("SCINE_BUILD_PYTHON_BINDINGS", "python"),
```

**Review by alalazo — CHANGES_REQUESTED** [review]
LGTM. A few minor requests

File: var/spack/repos/builtin/packages/serenity/package.py (hunk 4 above)
Comment (awvwgk) [reply var/spack/repos/builtin/packages/serenity/package.py:120]: The SERENITY_MARCH / SCINE_MARCH argument will set some march option unless it is an empty string.

Comment (alalazo) [reply var/spack/repos/builtin/packages/serenity/package.py:120]: Thanks for clarifying, all good then.

**Review by alalazo — APPROVED** [review]

**Pull request merged**